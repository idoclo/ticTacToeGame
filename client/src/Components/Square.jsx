import React from 'react';
import { Icon } from 'semantic-ui-react';
import PropTypes from 'prop-types';


const Square = ({ value, index, clickMethod, activePlayer, toggleActivePlayer, gameId, winner}) => {
  const handleClick = (i) => {
    if (!gameId) {
      window.alert('The game is over. Please start a new game.');
    } else if (value) {
      window.alert('Square occupied - please pick another square.');
    } else {
      const piece = activePlayer === 'playerX' ? 'X' : 'O';
      clickMethod(i, piece);
      // update active player piece
      const newActivePlayer = piece === 'X' ? 'playerO' : 'playerX';
      toggleActivePlayer(newActivePlayer);
    }
  }
  
  let icon;
  if (value === 'X') {
    icon = <Icon name="remove" size="huge" color="red"/>;
  } else if (value === 'O') {
    icon = <Icon name="radio" size ="huge" color="violet"/>;
  } else {
    icon = null;
  }

  return (
    <div
      className={winner ? "winning-square" : "square"}
      role="presentation"
      onClick={() => handleClick(index)}
    >
      {icon}
    </div>
  )
};


Square.propTypes = {
  value: PropTypes.string.isRequired,
  index: PropTypes.number.isRequired,
  clickMethod: PropTypes.func.isRequired,
  activePlayer: PropTypes.string.isRequired,
  toggleActivePlayer: PropTypes.func.isRequired,
  gameId: PropTypes.number.isRequired
};

export default Square;