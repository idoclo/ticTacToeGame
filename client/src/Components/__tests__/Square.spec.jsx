import React from 'react';
import { shallow } from 'enzyme';
import Square from '../Square';


describe('Square', () => {
  const shallowComponent = shallow(<Square value="X" index={0} clickMethod={() => {}} activePlayer="playerX" toggleActivePlayer={() => {}} gameId={10} winner={false}/>);

  it('renders correctly', () => {
    expect(shallowComponent).toMatchSnapshot();
  });

  it('calls handleClick when this Square is clicked on', () => {
    shallowComponent.instance().handleClick = jest.fn();
    const { handleClick } = shallowComponent.instance();
    expect(handleClick).toHaveBeenCalledTimes(0);
    shallowComponent.simulate('click', 1);
    expect(handleClick).toHaveBeenCalledTimes(1);
  });
});