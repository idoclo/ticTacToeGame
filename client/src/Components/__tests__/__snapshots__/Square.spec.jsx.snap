// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Square renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Square
    activePlayer="playerX"
    clickMethod={[Function]}
    gameId={10}
    index={0}
    toggleActivePlayer={[Function]}
    value="X"
    winner={false}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Icon
          as="i"
          color="red"
          name="remove"
          size="huge"
        />,
        <Portal
          closeOnDocumentClick={true}
          closeOnEscape={true}
          eventPool="default"
          onClose={[Function]}
          open={false}
          openOnTriggerClick={true}
        >
          <Segment
            style={
              Object {
                "left": "35%",
                "position": "fixed",
                "top": "40%",
                "zIndex": 1000,
              }
            }
          >
            <Header>
              <Icon
                as="i"
                name="ban"
                style={
                  Object {
                    "color": "red",
                  }
                }
              />
              Square occupied - please pick another square.
            </Header>
          </Segment>
        </Portal>,
      ],
      "className": "square",
      "onClick": [Function],
      "role": "presentation",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "as": "i",
          "color": "red",
          "name": "remove",
          "size": "huge",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <Segment
            style={
              Object {
                "left": "35%",
                "position": "fixed",
                "top": "40%",
                "zIndex": 1000,
              }
            }
          >
            <Header>
              <Icon
                as="i"
                name="ban"
                style={
                  Object {
                    "color": "red",
                  }
                }
              />
              Square occupied - please pick another square.
            </Header>
          </Segment>,
          "closeOnDocumentClick": true,
          "closeOnEscape": true,
          "eventPool": "default",
          "onClose": [Function],
          "open": false,
          "openOnTriggerClick": true,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <Header>
              <Icon
                as="i"
                name="ban"
                style={
                  Object {
                    "color": "red",
                  }
                }
              />
              Square occupied - please pick another square.
            </Header>,
            "style": Object {
              "left": "35%",
              "position": "fixed",
              "top": "40%",
              "zIndex": 1000,
            },
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": Array [
                <Icon
                  as="i"
                  name="ban"
                  style={
                    Object {
                      "color": "red",
                    }
                  }
                />,
                "Square occupied - please pick another square.",
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "as": "i",
                  "name": "ban",
                  "style": Object {
                    "color": "red",
                  },
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "Square occupied - please pick another square.",
            ],
            "type": [Function],
          },
          "type": [Function],
        },
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Icon
            as="i"
            color="red"
            name="remove"
            size="huge"
          />,
          <Portal
            closeOnDocumentClick={true}
            closeOnEscape={true}
            eventPool="default"
            onClose={[Function]}
            open={false}
            openOnTriggerClick={true}
          >
            <Segment
              style={
                Object {
                  "left": "35%",
                  "position": "fixed",
                  "top": "40%",
                  "zIndex": 1000,
                }
              }
            >
              <Header>
                <Icon
                  as="i"
                  name="ban"
                  style={
                    Object {
                      "color": "red",
                    }
                  }
                />
                Square occupied - please pick another square.
              </Header>
            </Segment>
          </Portal>,
        ],
        "className": "square",
        "onClick": [Function],
        "role": "presentation",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "as": "i",
            "color": "red",
            "name": "remove",
            "size": "huge",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <Segment
              style={
                Object {
                  "left": "35%",
                  "position": "fixed",
                  "top": "40%",
                  "zIndex": 1000,
                }
              }
            >
              <Header>
                <Icon
                  as="i"
                  name="ban"
                  style={
                    Object {
                      "color": "red",
                    }
                  }
                />
                Square occupied - please pick another square.
              </Header>
            </Segment>,
            "closeOnDocumentClick": true,
            "closeOnEscape": true,
            "eventPool": "default",
            "onClose": [Function],
            "open": false,
            "openOnTriggerClick": true,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": <Header>
                <Icon
                  as="i"
                  name="ban"
                  style={
                    Object {
                      "color": "red",
                    }
                  }
                />
                Square occupied - please pick another square.
              </Header>,
              "style": Object {
                "left": "35%",
                "position": "fixed",
                "top": "40%",
                "zIndex": 1000,
              },
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": Array [
                  <Icon
                    as="i"
                    name="ban"
                    style={
                      Object {
                        "color": "red",
                      }
                    }
                  />,
                  "Square occupied - please pick another square.",
                ],
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "as": "i",
                    "name": "ban",
                    "style": Object {
                      "color": "red",
                    },
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                "Square occupied - please pick another square.",
              ],
              "type": [Function],
            },
            "type": [Function],
          },
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
